### PR TITLE
Infinite list in action

### PR DESCRIPTION
**Pull request by krawieck**
infinite lists!!!

pros:
* sortable
* prependable

cons:
* state **does not** get preserved after switching tabs :(

### CODE REVIEW

File: lib/pages/communities_list.dart [hunk 1]
```diff
@@ -8,55 +9,65 @@ import '../widgets/markdown_text.dart';
...
-                  )
-                : SizedBox(width: 50),
-            // TODO: add trailing button for un/subscribing to communities
-          ),
-          itemCount: communities.length,
-        ));
+                      errorWidget: (_, __, ___) => SizedBox(width: 50),
+                    )
+                  : SizedBox(width: 50),
+              // TODO: add trailing button for un/subscribing to communities
```
Comment (shilangyu) [inline lib/pages/communities_list.dart:66]: I think we decided to remove this feature

File: lib/pages/community.dart [hunk 2]
```diff
@@ -207,16 +208,28 @@ class CommunityPage extends HookWidget {
...
-                  Center(child: Text('comments go here')),
-                ],
+              InfinitePostList(
+                fetcher: (page, batchSize, sort) =>
+                    LemmyApi(community.instanceUrl).v1.getPosts(
+                          type: PostListingType.community,
+                          sort: sort,
+                          communityId: community.id,
+                          page: page,
+                          limit: 10,
```
Comment (shilangyu) [inline lib/pages/community.dart:218]: set limit to batchSize

File: lib/pages/communities_list.dart [hunk 3]
```diff
@@ -8,55 +9,65 @@ import '../widgets/markdown_text.dart';
...
-                    width: 50,
-                    imageUrl: communities[i].icon,
-                    imageBuilder: (context, imageProvider) => Container(
-                      decoration: BoxDecoration(
-                        shape: BoxShape.circle,
-                        image: DecorationImage(
-                            fit: BoxFit.cover, image: imageProvider),
+      appBar: AppBar(
+        brightness: theme.brightness,
+        title: Text(title ?? '', style: theme.textTheme.headline6),
```
Comment (shilangyu) [inline lib/pages/communities_list.dart:28]: Set default instead of null coalescing

File: lib/widgets/infinite_scroll.dart [hunk 4]
```diff
@@ -54,9 +58,15 @@ class InfiniteScroll<T> extends HookWidget {
...
 
     return ListView.builder(
-      // +1 for the loading widget
-      itemCount: data.value.length + 1,
+      padding: padding,
+      // +2 for the loading widget and prepend widget
+      itemCount: data.value.length + 2,
       itemBuilder: (_, i) {
+        if (i == 0) {
+          return prepend ?? Container();
```
Comment (shilangyu) [inline lib/widgets/infinite_scroll.dart:66]: Set default instead of null coalescing

File: lib/widgets/post_list_options.dart [hunk 5]
```diff
@@ -57,17 +57,18 @@ class PostListOptions extends HookWidget {
...
           ),
           Spacer(),
-          IconButton(
-            icon: Icon(Icons.view_stream),
-            onPressed: () => print('TBD'),
-          )
+          if (styleButton)
+            IconButton(
+              icon: Icon(Icons.view_stream),
+              onPressed: () => print('TBD'),
```
Comment (shilangyu) [inline lib/widgets/post_list_options.dart:70]: add a TODO

File: lib/widgets/sortable_infinite_list.dart [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      batchSize: 20,
+    );
+  }
+}
+
+class InfinitePostList extends StatelessWidget {
+  final Future<List<PostView>> Function(
+      int page, int batchSize, SortType sortType) fetcher;
+
+  InfinitePostList({this.fetcher});
```
Comment (shilangyu) [inline lib/widgets/sortable_infinite_list.dart:55]: mark as required

File: lib/widgets/sortable_infinite_list.dart [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      fetcher: fetcher,
+    );
+  }
+}
+
+class InfiniteCommentList extends StatelessWidget {
+  final Future<List<CommentView>> Function(
+      int page, int batchSize, SortType sortType) fetcher;
+
+  InfiniteCommentList({this.fetcher});
```
Comment (shilangyu) [inline lib/widgets/sortable_infinite_list.dart:73]: mark as required

**Review by shilangyu — CHANGES_REQUESTED** [review]

File: lib/widgets/infinite_scroll.dart (hunk 4 above)
Comment (krawieck) [reply lib/widgets/infinite_scroll.dart:66]: image
can't do unfortunately :(

Comment (shilangyu) [reply lib/widgets/infinite_scroll.dart:66]: const Container()

**Review by shilangyu — APPROVED** [review]

**Pull request merged**